### ISSUE TITLE
Reduce log spam on missing member

### ISSUE THREAD

**Issue by davhdavh** [etcd-io/etcd#18238]
### What would you like to be added?

```
{"level":"warn","ts":"2024-06-27T07:54:41.820665Z","caller":"rafthttp/peer.go:267","msg":"dropped internal Raft message since sending buffer is full (overloaded network)","message-type":"MsgHeartbeat","local-member-id":"e91c799c9231b7d2","from":"e91c799c9231b7d2","remote-peer-id":"d17db337c2371a6d","remote-peer-name":"pipeline","remote-peer-active":false}
{"level":"warn","ts":"2024-06-27T07:54:41.831868Z","caller":"rafthttp/peer.go:267","msg":"dropped internal Raft message since sending buffer is full (overloaded network)","message-type":"MsgHeartbeat","local-member-id":"e91c799c9231b7d2","from":"e91c799c9231b7d2","remote-peer-id":"d17db337c2371a6d","remote-peer-name":"pipeline","remote-peer-active":false}
{"level":"warn","ts":"2024-06-27T07:54:41.831966Z","caller":"rafthttp/peer.go:267","msg":"dropped internal Raft message since sending buffer is full (overloaded network)","message-type":"MsgHeartbeat","local-member-id":"e91c799c9231b7d2","from":"e91c799c9231b7d2","remote-peer-id":"d17db337c2371a6d","remote-peer-name":"pipeline","remote-peer-active":false}
{"level":"warn","ts":"2024-06-27T07:54:41.931594Z","caller":"rafthttp/peer.go:267","msg":"dropped internal Raft message since sending buffer is full (overloaded network)","message-type":"MsgHeartbeat","local-member-id":"e91c799c9231b7d2","from":"e91c799c9231b7d2","remote-peer-id":"d17db337c2371a6d","remote-peer-name":"pipeline","remote-peer-active":false}
{"level":"warn","ts":"2024-06-27T07:54:41.981508Z","caller":"rafthttp/peer.go:267","msg":"dropped internal Raft message since sending buffer is full (overloaded network)","message-type":"MsgHeartbeat","local-member-id":"e91c799c9231b7d2","from":"e91c799c9231b7d2","remote-peer-id":"d17db337c2371a6d","remote-peer-name":"pipeline","remote-peer-active":false}
{"level":"warn","ts":"2024-06-27T07:54:42.130708Z","caller":"rafthttp/peer.go:267","msg":"dropped internal Raft message since sending buffer is full (overloaded network)","message-type":"MsgHeartbeat","local-member-id":"e91c799c9231b7d2","from":"e91c799c9231b7d2","remote-peer-id":"d17db337c2371a6d","remote-peer-name":"pipeline","remote-peer-active":false}
{"level":"warn","ts":"2024-06-27T07:54:42.13223Z","caller":"rafthttp/peer.go:267","msg":"dropped internal Raft message since sending buffer is full (overloaded network)","message-type":"MsgHeartbeat","local-member-id":"e91c799c9231b7d2","from":"e91c799c9231b7d2","remote-peer-id":"d17db337c2371a6d","remote-peer-name":"pipeline","remote-peer-active":false}
```

im fairly certain that giving me the exact same log message hundreds of times per seconds provide no actual benefit.

### Why is this needed?

because having to write 10s of megabytes per min of log on every single control-plane while a member is temporarily missing is insane.

**Comment by jmhbnz** [comment]
> Hmm, guessing without looking up documentation, do we have command lines for configuring log sampling in zap?

Unless I am missing something there is nothing in server logging config currently for zap sampling rates: https://etcd.io/docs/v3.6/op-guide/configuration/#logging

```bash
--logger 'zap'
  Currently only supports 'zap' for structured logging.
--log-outputs 'default'
  Specify 'stdout' or 'stderr' to skip journald logging even when running under systemd, or list of comma separated output targets.
--log-level 'info'
  Configures log level. Only supports debug, info, warn, error, panic, or fatal.
--enable-log-rotation 'false'
  Enable log rotation of a single log-outputs file target.
--log-rotation-config-json '{"maxsize": 100, "maxage": 0, "maxbackups": 0, "localtime": false, "compress": false}'
  Configures log rotation if enabled with a JSON logger config. MaxSize(MB), MaxAge(days,0=no limit), MaxBackups(0=no limit), LocalTime(use computers local time), Compress(gzip)".

```

We could consider adding https://pkg.go.dev/go.uber.org/zap#SamplingConfig for `rafthttp`?

**Comment by serathius** [comment]
~ I would be careful about overcomplicating logging, my question came from idea whether it can be solved now. If we have a problem with a single log line, we should look into it first, before we start supporting a new feature.

**Comment by moficodes** [comment]
Discussed during bug triage. Could be a good first issue with a bit more digging. @serathius @jmhbnz

**Comment by jmhbnz** [comment]
The issue is in rafthttp here: https://github.com/etcd-io/etcd/blob/3aa56a76bd7cda7a573bdc858af705810466dfb6/server/etcdserver/api/rafthttp/remote.go#L61 we need to reproduce the issue initially. Changing from feature to bug as we should expect log messages to not be spammy.

Thanks @davhdavh for reporting this, was there anything specific in your setup to reproduce this?

**Comment by mello369** [comment]
Hi everyone, I'd like to work on this issue, but would need some instructions on how I can reproduce it. Thanks.

**Comment by davhdavh** [comment]
Nothing special. Just removed a control plane node from k8s, but forgot to remove from etcd.

**Comment by jmhbnz** [comment]
Thanks for confirming @davhdavh and volunteering to take a look at it @mello369!

/assign @mello369

**Comment by jmhbnz** [comment]
Discussed during sig-etcd triage meeting, @mello369 how are you going with this one?  Did you need help with anything?

**Comment by NavinShrinivas** [comment]
Just to keep y'all in the loop, I have been able to replicate this issue in a minikube ha cluster with 3 nodes. Will look into the issue more. Do let me know if you want help replicating this as well @mello369

**Comment by mello369** [comment]
@NavinShrinivas yes, i would need help replicating this. Thanks.
@jmhbnz sorry for the inactivity on the thread, ill try to expedite this over the weekend.

**Comment by NavinShrinivas** [comment]
I've been rummaging through the code, and came across this : 
https://github.com/etcd-io/etcd/blob/68e712276bb8a8f35f7f077add143b28422e8d66/server/etcdserver/api/rafthttp/peer.go#L337-L348

And also read that pipelining is used only when streams are not ready. From the logs, we can see that this is coming from "pipeline". 

I would like some help as to in what possible cases could the stream not be initialised

**Comment by NavinShrinivas** [comment]
I spent some time look at this issue today and here are my findings, this spam is due to the leader attempting to send out heartbeat messages to the missing peer. We don't want to actually stop the retries (In the event that the node comes back on)

https://github.com/etcd-io/etcd/blob/68e712276bb8a8f35f7f077add143b28422e8d66/server/etcdserver/api/rafthttp/util.go#L47-L53

This is the round-tripper that is being used by the leader in the given logs. As it (by design) doesn't have a request/response time out. 
If a `hearbeat` was sent, and the node was abruptly stopped. The function that uses this round tripper: 
https://github.com/etcd-io/etcd/blob/68e712276bb8a8f35f7f077add143b28422e8d66/server/etcdserver/api/rafthttp/pipeline.go#L133-L155

Will indefinitely wait.

Causing the `msgc` buffer to fill up, as nothing is being consumed by 
https://github.com/etcd-io/etcd/blob/68e712276bb8a8f35f7f077add143b28422e8d66/server/etcdserver/api/rafthttp/pipeline.go#L93-L106 

as this is also blocked by the waiting `post` function. The filled `msgc` buffer is causing the spam of logs. 

As for solutions: 

1. Add a sensible request/response timeout 
2. Make the post function call from handle function a concurrent one, But I cannot judge of the correctness of this solution. 

Do let me know what you guys think.

**Comment by serathius** [comment]
I think the problem was spam of logs? What you propose with timeout and concurrency might be also good, but sounds like a overkill change just to reduce number of generated logs. Changing timeout here sounds like a much riskier change than needed. If you think it should addressed independently, please file another issue with suggestion.

**Comment by NavinShrinivas** [comment]
How about not logging (or logging it a few times) when we try and pipeline `hearbeat` messages in the msgc buffer and it fails?
